dvc/rwlock.py

```diff
@@ -61,38 +60,25 @@ def _infos_to_str(infos):
     )
 
 
-def _check_no_writers(lock, info, path_infos):
-    for path_info in path_infos:
-        blocking_urls = lfilter(path_info.overlaps, lock["write"])
-        if not blocking_urls:
-            continue
+def _check_blockers(lock, info, *, mode, waiters):
+    for path_info in waiters:
+        blockers = [
+            info
+            for path, infos in lock[mode].items()
+            if path_info.overlaps(path)
+            if info not in (infos if type(infos) is list else [infos])
+        ]
 
-        writers = lmap(lock["write"].get, blocking_urls)
-        writers = lfilter(lambda i: info != i, writers)
-        if not writers:
+        if not blockers:
             continue
 
         raise LockError(
-            "'{}' is busy, it is being written to by:\n{}".format(
-                str(path_info), _infos_to_str(writers)
-            )
-        )
-
-
-def _check_no_readers(lock, info, path_infos):
-    for path_info in path_infos:
-        blocking_urls = lfilter(path_info.overlaps, lock["read"])
-        if not blocking_urls:
-            continue
-
-        readers = lmapcat(lock["read"].get, blocking_urls)
-        readers = lfilter(lambda i: info != i, readers)
-        if not readers:
-            continue
-
-        raise LockError(
-            "'{}' is busy, it is being read by:\n{}".format(
-                str(path_info), _infos_to_str(readers)
+            "'{path}' is busy, it is being blocked by:\n"
+            "{blockers}\n"
```

Comment:
is it an array now? why? it should be always single blocker, right? let's put an assert and print the first element?

Reply:
@shcheklein , it has been always like that, it is more clear now :slightly_smiling_face: 

This implementation of RWLock allows multiple readers to a single file

Reply:
ok, I see. it's quite strange by itself, but not related to this PR 👍

Reply:
what's strange, @shcheklein ? :slightly_smiling_face:

Reply:
that blockers is an array ... unless I'm missing something there should 0 or 1 _blocker_

Reply:
@shcheklein , blockers contain process information, there could be many processes reading the same file, those processes are blockers for a writer.

hope that brings some clarity

Reply:
my bad, you are right ... multiple readers situation is possible .. ideally we should take this into account in the message.

Reply:
I mean put process into plural, ID into IDs ... etc ... probably minor nuance, but can be confusing a bit for those who read code

Reply:
done, @shcheklein )

Reply:
thanks @mroutis !